From 845974d62b5d2b6acfd312107daff527a3cc611d Mon Sep 17 00:00:00 2001
From: "yanbin.fang" <yanbin.fang@spreadtrum.com>
Date: Mon, 28 Apr 2014 13:36:57 +0800
Subject: [PATCH] Bug #304668 can't get semphore in get_userprocess_meminfo
 sometimes.

[bug number  ]
[root cause  ]
[changes     ]
[side effects]
[self test   ] test
[reviewers   ]

Change-Id: I0c42fb0c02bfb496e3cd7d1ab163d3421efb4632
---
 drivers/staging/android/lowmemorykiller.c |    2 +-
 fs/proc/proc_usermeminfo.c                |    6 ++++--
 mm/oom_kill.c                             |    2 +-
 3 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/android/lowmemorykiller.c b/drivers/staging/android/lowmemorykiller.c
index 80f5036..497f917 100644
--- a/drivers/staging/android/lowmemorykiller.c
+++ b/drivers/staging/android/lowmemorykiller.c
@@ -283,7 +283,7 @@ static int lowmem_shrink(struct shrinker *s, struct shrink_control *sc)
 #ifdef CONFIG_ANDROID_LMK_DEBUG
 		if (selected_oom_adj <= 16) {
 			dump_header(current, sc->gfp_mask, -1, 0, 0);
-			//user_process_meminfo_show();
+			user_process_meminfo_show();
 #ifdef CONFIG_ZRAM
 			zram_printlog();
 #endif
diff --git a/fs/proc/proc_usermeminfo.c b/fs/proc/proc_usermeminfo.c
index 9926350..e165241 100644
--- a/fs/proc/proc_usermeminfo.c
+++ b/fs/proc/proc_usermeminfo.c
@@ -166,6 +166,7 @@ int user_process_meminfo_show(void)
 			"stat", "pid", "vss", "rss", "rss_mapped", "pss", "uss", "swap");
 
 	printk(KERN_INFO "-----------------------------------------------------------------\n");
+	read_lock(&tasklist_lock);
 	for_each_process(p) {
 		memset(&upmeminfo, 0, sizeof(upmeminfo));
 		if (get_userprocess_meminfo(p, &upmeminfo))
@@ -182,7 +183,7 @@ int user_process_meminfo_show(void)
 				mm_counter << (PAGE_SHIFT - 10),
 				p->comm);
 	}
-
+	read_unlock(&tasklist_lock);
 	return 0;
 }
 
@@ -195,6 +196,7 @@ static int proc_user_process_meminfo_show(struct seq_file *m, void *v)
 			"pid", "vss", "rss", "rss_mapped", "pss", "uss", "name");
 
 	seq_printf(m, "-----------------------------------------------------------------\n");
+	read_lock(&tasklist_lock);
 	for_each_process(p) {
 		memset(&upmeminfo, 0, sizeof(upmeminfo));
 		if (get_userprocess_meminfo(p, &upmeminfo))
@@ -209,7 +211,7 @@ static int proc_user_process_meminfo_show(struct seq_file *m, void *v)
 				upmeminfo.uss >> 10,
 				p->comm);
 	}
-
+	read_unlock(&tasklist_lock);
 	return 0;
 }
 
diff --git a/mm/oom_kill.c b/mm/oom_kill.c
index 474038c..be14b12 100644
--- a/mm/oom_kill.c
+++ b/mm/oom_kill.c
@@ -425,7 +425,7 @@ static void dump_header(struct task_struct *p, gfp_t gfp_mask, int order,
 #ifdef CONFIG_ZRAM
 		zram_printlog();
 #endif
-		//user_process_meminfo_show();
+		user_process_meminfo_show();
 #else
 		dump_tasks(mem, nodemask);
 #endif
-- 
1.7.9.5

